src/components/PostDetails.tsx

```diff
@@ -1,90 +1,67 @@
-
 import * as commentsApi from '../api/comments';
-
+import { CommentData } from '../types/Comment';
+import { useAppSelector } from '../hooks/useAppSelector.hook';
+import {
+  addNewComment,
+  fetchPostComments,
+  takeError,
+  deleteComment as deleteChosenComment,
+} from '../slices/comments.slice';
+import { useAppDispatch } from '../hooks/useAppDispatch.hook';
 import { Post } from '../types/Post';
-import { Comment, CommentData } from '../types/Comment';
 
 type Props = {
   post: Post;
 };
 
 export const PostDetails: React.FC<Props> = ({ post }) => {
-  const [comments, setComments] = useState<Comment[]>([]);
-  const [loaded, setLoaded] = useState(false);
-  const [hasError, setError] = useState(false);
+  const dispatch = useAppDispatch();
+  const {
+    loaded,
+    hasError,
+    items: comments,
+  } = useAppSelector(state => state.comments);
   const [visible, setVisible] = useState(false);
 
-  function loadComments() {
-    setLoaded(false);
-    setError(false);
-    setVisible(false);
-
-    commentsApi
-      .getPostComments(post.id)
-      .then(setComments) // save the loaded comments
-      .catch(() => setError(true)) // show an error when something went wrong
-      .finally(() => setLoaded(true)); // hide the spinner
-  }
-
-  useEffect(loadComments, [post.id]);
-
   // The same useEffect with async/await
-  /*
-  async function loadComments() {
-    setLoaded(false);
-    setVisible(false);
-    setError(false);
-
-    try {
-      const commentsFromServer = await commentsApi.getPostComments(post.id);
 
-      setComments(commentsFromServer);
-    } catch (error) {
-      setError(true);
-    } finally {
-      setLoaded(true);
-    }
+  const loadComments = (newPostId: number) => {
+    dispatch(fetchPostComments(newPostId));
   };
 
   useEffect(() => {
-    loadComments();
-  }, []);
+    setVisible(false);
+    loadComments(post.id);
+  }, [post]);
 
-  useEffect(loadComments, [post.id]); // Wrong!
   // effect can return only a function but not a Promise
-  */
 
   const addComment = async ({ name, email, body }: CommentData) => {
-    try {
-      const newComment = await commentsApi.createComment({
-        name,
-        email,
-        body,
-        postId: post.id,
-      });
-
-      setComments(currentComments => [...currentComments, newComment]);
-
-      // setComments([...comments, newComment]);
-      // works wrong if we wrap `addComment` with `useCallback`
-      // because it takes the `comments` cached during the first render
-      // not the actual ones
-    } catch (error) {
-      // we show an error message in case of any error
-      setError(true);
+    if (post) {
+      try {
+        const newComment = await commentsApi.createComment({
+          name,
+          email,
+          body,
+          postId: post?.id,
+        });
+
+        dispatch(addNewComment(newComment));
+      } catch (error) {
+        // we show an error message in case of any error
+        takeError();
```

Comment:
The function `takeError` should be dispatched to update the state with the error. Use `dispatch(takeError())` instead.